File: JVM/src/main/java/io/alliance/adn/Walkable.java
package io.alliance.adn;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;

@RequiredArgsConstructor
public abstract class Walkable<T> {
    @NotNull
    protected final T[] input;
    protected int index = 0;

    @NotNull
    protected T current() {
        return input[index];
    }

    @NotNull
    protected T next() {
        return input[index + 1];
    }

    @NotNull
    protected T peek(int offset) {
        return input[index + offset];
    }

    @NotNull
    protected T consume() {
        final T result = input[index];
        index++;
        return result;
    }

    protected void skip() {
        index++;
    }
}

File: JVM/src/main/kotlin/io/alliance/adn/Walkable.kt
package io.alliance.adn

internal abstract class Walkable<T> protected constructor(
    @JvmField protected val input: List<T>,
    @JvmField protected var index: Int = 0
) {
    protected val size: Int get() = input.size

    protected val current: T get() = input[index]

    protected val next: T get() = input[index + 1]

    protected fun peek(offset: Int): T {
        return input[index + offset]
    }

    protected val consume: T get() {
        val result = input[index]
        index++
        return result
    }

    protected fun skip() {
        index++
    }
}